### PR TITLE
Added LeaderScheduler module and tests

### PR DESCRIPTION
**Pull request by carllin**

### CODE REVIEW

File: src/hash.rs [hunk 1]
```diff
@@ -7,7 +7,7 @@ use sha2::{Digest, Sha256};
 use std::fmt;
 
 #[derive(Serialize, Deserialize, Clone, Copy, Default, Eq, PartialEq, Ord, PartialOrd, Hash)]
-pub struct Hash(GenericArray<u8, U32>);
+pub struct Hash(pub GenericArray<u8, U32>);
```
Comment (garious) [inline src/hash.rs]: What adds a dependency on the internal type? I was hoping to change this to a fixed-size array at some point?

**Comment by garious** [comment]
This will take me some time to consume, but looks very nice!

File: src/hash.rs (hunk 1 above)
Comment (carllin) [reply src/hash.rs]: That would work as well! It's consumed by the calculate_seed() function which needs access to the byte array of the hash.

**Comment by garious** [comment]
```
$ cargo +nightly install cargo-cov
$ cargo +nightly cov test --lib
$ cargo +nightly report --open
...
/src/leader_scheduler.rs | 96.31% | 36.49%
```

Your test suite is much better than average in the solana codebase, but it consistently misses the edge cases. Try running cargo-cov locally and see if you can either remove some of those edge cases or write tests for them.

Other than that, this PR is just amazing. Adding tests for those edge cases would only turn this A work into A+ work. :)

**Comment by carllin** [comment]
That's an awesome tool I didn't know about, thanks for the tip and the review!

**Comment by garious** [comment]
Doh, try not to squash commits after we've started reviewing. I can't tell what changed.

so... what changed?

**Comment by carllin** [comment]
Sorry, I think I just rebased my changes yesterday.

**Comment by carllin** [comment]
Also, I think the test is failing because the tpu now generates ticks, so it's no longer valid to assume that after shutting down the tvu and restarting the tpu in handle_role_transition() that the ledger will be unchanged. I'll add an ignore to that test, as I'll be revamping it in my next PR for the plumbing of this LeaderScheduler anyways.

**Comment by garious** [comment]
cc #1081

File: src/leader_scheduler.rs [hunk 2]
```diff
@@ -0,0 +1,776 @@
...
+        self.active_validators.get_active_set(height)
+    }
+
+    pub fn update_height(&mut self, height: u64, bank: &Bank) {
+        if height < self.bootstrap_height {
+            return;
+        }
+
+        if height == self.bootstrap_height
+            || (height - self.bootstrap_height) % self.seed_rotation_interval == 0
```
Comment (rob-solana) [inline src/leader_scheduler.rs]: the first clause is redundant?

File: src/leader_scheduler.rs [hunk 3]
```diff
@@ -0,0 +1,776 @@
...
+            seed_rotation_interval,
+            leader_schedule: Vec::new(),
+            last_seed_height: None,
+            bootstrap_leader: config.bootstrap_leader,
+            bootstrap_height,
+            seed: 0,
+        }
+    }
+
+    pub fn push_vote(&mut self, id: Pubkey, height: u64) {
```
Comment (rob-solana) [inline src/leader_scheduler.rs]: is there a case where push_vote() does not imply update_height()?

Comment (rob-solana) [reply src/leader_scheduler.rs]: I guess we want to keep these separate so that there's just one place where we update height...

File: src/leader_scheduler.rs (hunk 2 above)
Comment (carllin) [reply src/leader_scheduler.rs]: I think you can get overflow errors if height < bootstrap height when calculating height - self.bootstrap_height in the second clause.

Comment (rob-solana) [reply src/leader_scheduler.rs]: I mean, if they're equal, height - self.bootstrap_height % anything will be zero...

Comment (carllin) [reply src/leader_scheduler.rs]: oh doh, yeah xD
